New Adapter: Scalibur

bTimor opened the pull request:
## Type of change
- [x] New bidder adapter

## Description
This PR adds the Scalibur bidder adapter to Prebid Server.

### Features:
- Supports Banner and Video media types.
- Implements User Sync via iframe.
- Automated tests include standard unit tests and JSON data-driven samples.

### Verification:
- End-to-end user sync flow verified using a private Scalibur sync server. 
- Successfully verified that PBS maps the `uids` cookie to `user.buyeruid` in outgoing requests.
- All tests in `adapters/scalibur` passed.
- Root `./validate.sh` passed (ignoring unrelated global config/hook test timeouts).

## Checklist
- [x] I have performed a self-review of my code
- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works

https://github.com/prebid/prebid.github.io/pull/6381

Commit:
fresh

Commit:
fix ortb and sync server support

Commit:
finish scalibur prebid adapter

Commit:
Merge remote-tracking branch 'upstream/master' into scaliburA

Commit:
Add Scalibur bid adapter with Banner, Video, and User Sync support

Commit:
add ext debug

Commit:
add gzip compression support

Commit:
resolve multi type response

ccorbo commented on adapters/scalibur/scalibur.go:
```diff
@@ -0,0 +1,316 @@
+			impCopy.BidFloorCur = scaliburExt.BidFloorCur
+		}
+		if impCopy.BidFloorCur == "" {
+			impCopy.BidFloorCur = "USD"
```
question: could there be currency conversion issues if the bid floor isn't usd but you are stamping as usd bid floor?
ExtraRequestInfo has currency converter functionality

ccorbo commented on adapters/scalibur/scalibur.go:
```diff
@@ -0,0 +1,316 @@
+func getBidMediaType(bid openrtb2.Bid, imp *openrtb2.Imp) (openrtb_ext.BidType, error) {
+	// 1 = Banner, 2 = Video, 3 = Audio, 4 = Native
+	switch bid.MType {
+	case 1:
```
avoid magic numbers and use constants from openrtb2 package
```
MarkupBanner MarkupType = 1
MarkupVideo  MarkupType = 2
MarkupAudio  MarkupType = 3
MarkupNative MarkupType = 4
```

ccorbo commented on static/bidder-params/scalibur.json:
```diff
@@ -0,0 +1,22 @@
+{
```
add params test file to cover this. Example: https://github.com/prebid/prebid-server/blob/master/adapters/appnexus/params_test.go

Commit:
address PR review feedback and fix bidder implementation

Commit:
Merge branch 'master' into scaliburA

ccorbo commented on adapters/scalibur/scalibur.go:
```diff
@@ -0,0 +1,330 @@
+	isDebug := request.Test == 1
+	if !isDebug && len(request.Ext) > 0 {
+		var extRequest openrtb_ext.ExtRequest
+		if err := jsonutil.Unmarshal(request.Ext, &extRequest); err == nil {
```
add test coverage via json tests here

ccorbo commented on static/bidder-info/scalibur.yaml:
```diff
@@ -0,0 +1,34 @@
+endpoint: "https://srv.scalibur.io/adserver/ortb?type=prebid-server"
```
confirmed endpoint:

<img width="814" height="259" alt="image" src="https://github.com/user-attachments/assets/2e9a847a-6a6d-4f15-b4e0-af3fe9f831f3" />

Commit:
address PR feedback for Scalibur adapter

ccorbo reviewed (commented):
couple small things but overall lgtm, will approve after next changes

Commit:
fix(scalibur): address additional PR feedback

- Remove unsupported audio/native types from getBidMediaType
- Add JSON test for video mtype
- Add unit tests for parseScaliburExt error cases

Commit:
Merge branch 'master' into scaliburA

Commit:
Merge branch 'master' into scaliburA

Commit:
Merge branch 'master' into scaliburA

Commit:
Merge remote-tracking branch 'upstream/master' into scaliburA

Commit:
Merge branch 'master' into scaliburA

Commit:
merge from master

danylo-kazymyrov-teqblaze commented on exchange/adapter_builders.go:
```diff
@@ -471,6 +473,8 @@ func newAdapterBuilders() map[openrtb_ext.BidderName]adapters.Builder {
 		openrtb_ext.BidderSeedtag:           seedtag.Builder,
 		openrtb_ext.BidderSaLunaMedia:       salunamedia.Builder,
+		openrtb_ext.BidderScalibur:          scalibur.Builder,
+		openrtb_ext.BidderScreencore:        screencore.Builder,
```
@bTimor this is unexpected changes. You should not modify adapters except your. Pls revert Screencore changes

danylo-kazymyrov-teqblaze reviewed (commented):
left comments

Commit:
Merge branch 'master' into scaliburA

Commit:
address pr comments

bTimor committed:
Merge branch 'master' into scaliburA

bTimor committed:
Merge branch 'master' into scaliburA

bTimor committed:
Merge branch 'master' into scaliburA

bTimor committed:
Merge branch 'master' into scaliburA

bTimor committed:
Merge branch 'master' into scaliburA

bTimor commented:
@bsardo , each time I update the branch with main, it asks me to re-validate the merge.

bTimor committed:
Merge branch 'master' into scaliburA

bTimor commented:
@bsardo , should I continue resolving the conflicts myself, or should I just wait for one of you to do it when you try to merge?

Commit:
merge

Commit:
Merge branch 'scaliburA' of github.com:scalibur-io/prebid-server into scaliburA

bTimor committed:
Merge branch 'master' into scaliburA

Commit:
update to v4

Commit:
Merge branch 'scaliburA' of github.com:scalibur-io/prebid-server into scaliburA

bTimor committed:
Merge branch 'master' into scaliburA

bTimor commented:
@danylo-kazymyrov-teqblaze, all over again. do I do it wrong?

danylo-kazymyrov-teqblaze commented:
@bsardo could you merge this Adapter?

bTimor committed:
Merge branch 'master' into scaliburA

bTimor commented:
@bsardo Can you please merge the new adapter?

bTimor committed:
Merge branch 'master' into scaliburA

bTimor commented:
@bsardo , @ccorbo  , @danylo-kazymyrov-teqblaze, I’ve been trying to merge my branch for a few months now.

Am I doing something wrong? Could you please merge it so I can finally start working with it?

bsardo commented on static/bidder-info/scalibur.yaml:
```diff
@@ -0,0 +1,34 @@
+  email: "support@scalibur.io"
+supported-media-types:
+  - banner
+  - video
```
Please remove these lines. `supported-media-types` is not a valid PBS-Go field.

bsardo commented on static/bidder-info/scalibur.yaml:
```diff
@@ -0,0 +1,34 @@
+endpoint: "https://srv.scalibur.io/adserver/ortb?type=prebid-server"
+
```
Please remove all of the blank lines in this file.

bsardo commented on docs/bidders/scalibur.md:
Please remove this file and instead create a PR in the docs repo [prebid.github.io](https://github.com/prebid/prebid.github.io) to add this information.

bsardo commented on adapters/scalibur/scalibur.go:
```diff
@@ -0,0 +1,326 @@
+		return nil, &errortypes.BadInput{
+			Message: "placementId is required",
+		}
+	}
```
You can delete this code. Your adapter will never be called without a placementId because you have declared it as required with a minimum length of 1 in `bidder-params/scalibur.json`.

bsardo commented on adapters/scalibur/scalibur.go:
```diff
@@ -0,0 +1,326 @@
+
+		impCopy := imp
+
+		// Apply bid floor and currency
+		if scaliburExt.BidFloor != nil && *scaliburExt.BidFloor > 0 {
+			impCopy.BidFloor = *scaliburExt.BidFloor
+			if scaliburExt.BidFloorCur != "" {
+				impCopy.BidFloorCur = scaliburExt.BidFloorCur
+			}
+		}
+
+		if impCopy.BidFloor > 0 && impCopy.BidFloorCur != "" && impCopy.BidFloorCur != "USD" {
+			convertedValue, err := reqInfo.ConvertCurrency(impCopy.BidFloor, impCopy.BidFloorCur, "USD")
+			if err != nil {
+				errs = append(errs, err)
+				continue
+			}
+			impCopy.BidFloor = convertedValue
+			impCopy.BidFloorCur = "USD"
+		}
```
Please add supplemental JSON test(s) to ensure coverage of these lines. They are mostly uncovered by the JSON framework which is required wherever possible.

bsardo reviewed (commented):
@bTimor sorry for the delay. Please see the comments I just posted. Once these are addressed we can merge.

Commit:
Merge branch 'master' into scaliburA

Commit:
fix all last pr comments and merge with master

Commit:
Merge branch 'scaliburA' of github.com:scalibur-io/prebid-server into scaliburA

Commit:
remove blank lines from scalibur yaml

bTimor commented:
@bsardo thanks for the response. I resolved all the requested issues and updated the branch with the last master.
Waiting for your approval and merge.

bTimor committed:
Merge branch 'master' into scaliburA

bTimor commented:
@bsardo , @ccorbo , @danylo-kazymyrov-teqblaze any update?

bTimor committed:
Merge branch 'master' into scaliburA

bTimor commented:
@bsardo , @ccorbo , @danylo-kazymyrov-teqblaze, @pm-jaydeep-mohite  this has become extremely urgent. I’ve spent a lot of extra time keeping the PR up to date and pinging you all, while no actual progress has been made. I now have a major client waiting on it.

Please let me know if there is someone else I should contact in order to get the PR merged.

bTimor commented:
@bsardo  , @ccorbo  , @danylo-kazymyrov-teqblaze , @pm-jaydeep-mohite , still needs 2 approvals and merge.

bsardo commented on adapters/scalibur/scalibur.out:
Please delete this file.

bTimor replied:
Done

bsardo reviewed (commented):
@bTimor address the one comment and I will approve and get this merged.

Commit:
Merge branch 'master' into scaliburA

Commit:
remove out file

Commit:
Merge branch 'scaliburA' of github.com:scalibur-io/prebid-server into scaliburA

bTimor commented:
@bsardo deleted the file.

bTimor commented:
@bsardo  , @ccorbo  , @danylo-kazymyrov-teqblaze  , @pm-jaydeep-mohite. Resolve all comments. waiting for the merge.
